rust/bridge/jni/bin/gen_java_decl.py

```diff
@@ -0,0 +1,152 @@
+    tempFile.deleteOnExit();
+
+    try (OutputStream out = new FileOutputStream(tempFile)) {
+      byte[] buffer = new byte[4096];
+      int read;
+
+      while ((read = in.read(buffer)) != -1) {
+        out.write(buffer, 0, read);
+      }
+    }
+    System.load(tempFile.getAbsolutePath());
+  }
+
+  /*
+  If a .so and/or .dylib is embedded within this jar as a resource file, attempt
+  to copy it to a temporary file and then load it. This allows the jar to be
+  used even without a libsignal_jni shared library existing on the filesystem.
+  */
+  private static void loadLibrary() {
+    try {
+      String  osName    = System.getProperty("os.name").toLowerCase(java.util.Locale.ROOT);
+      boolean isMacOs   = osName.startsWith("mac os");
+      String  extension = isMacOs ? ".dylib" : ".so";
+
+      try (InputStream in = Native.class.getResourceAsStream("/libsignal_jni" + extension)) {
+        if (in != null) {
+          copyToTempFileAndLoad(in, extension);
+        } else {
+          System.loadLibrary("signal_jni");
+        }
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  static {
+    loadLibrary();
+  }
+
+  private Native() {}
+"""
+
+def translate_to_java(typ):
+    if typ == "void":
+        return "void"
+    if typ == "jstring":
+        return "String"
+    if typ == "jbyteArray":
+        return "byte[]"
+    if typ == "ObjectHandle":
+        return "long"
+    if typ == "jint":
+        return "int"
+    if typ == "jlong":
+        return "long"
+    if typ == "jboolean":
+        return "boolean"
+    if typ == "JString":
+        return "String"
+    if typ == "jobject":
+        return "Object"
+
+    raise Exception("Don't know what to do with a", typ)
+
+cur_type = None
+
+for line in stdout.split('\n'):
+    if line == '':
+        continue
+
+    match = java_decl.match(line)
+    if match is None:
+        raise Exception("Could not understand", line)
+
+    if cur_type is None or match.group(2) != cur_type:
+        contents += "\n"
+        cur_type = match.group(2)
+    java_fn_name = '%s_%s' % (match.group(2), match.group(3))
+    java_ret_type = translate_to_java(match.group(1))
+    java_args = []
+
+    if match.group(4) != None:
+        args = match.group(4).split(', ')[1:]
+
+        for arg in args:
+            (arg_type,arg_name) = arg.split(' ')
+            java_arg_type = translate_to_java(arg_type)
+            java_args.append('%s %s' % (java_arg_type, arg_name))
+
+    contents += "  public static native %s %s(%s);\n" % (java_ret_type, java_fn_name, ", ".join(java_args))
+
+contents += "}\n"
+
+native_java = os.path.join(our_abs_dir, '../../../../java/java/src/main/java/org/signal/internal/Native.java')
+
+if not os.access(native_java, os.F_OK):
+    raise Exception("Didn't find Native.java where it was expected")
+
+fh = open(native_java, 'w')
+fh.write(contents)
+fh.close()
```

Comment:
If this does get moved into the build system, it'd be easier to write to stdout and not worry about the path on disk.